src/v/storage/ntp_config.h

```diff
@@ -15,18 +15,64 @@
 #include "model/metadata.h"
 #include "model/namespace.h"
 #include "ssx/sformat.h"
+#include "storage/file_sanitizer_types.h"
 #include "utils/tristate.h"
 
 #include <seastar/core/sstring.hh>
 
+#include <fmt/format.h>
+
 #include <optional>
+#include <ostream>
 
 namespace storage {
 using with_cache = ss::bool_class<struct log_cache_tag>;
 using topic_recovery_enabled
   = ss::bool_class<struct topic_recovery_enabled_tag>;
 
+/// This represents a set of capabilities that a topic can have.
+class ntp_caps {
+public:
+    explicit ntp_caps(
+      bool tiered_storage, bool iceberg, bool cloud_topics) noexcept
+      : _tiered_storage(tiered_storage)
+      , _iceberg(iceberg)
+      , _cloud_topics(cloud_topics) {}
+
+    /// Tiered storage can be enabled for this NTP
+    bool has_tiered_storage() const noexcept { return _tiered_storage; }
+    /// Iceberg can be enabled for this NTP
+    bool has_iceberg() const noexcept { return _iceberg; }
+    /// Cloud topics can be enabled for this NTP
+    bool has_cloud_topics() const noexcept { return _cloud_topics; }
+
+private:
+    bool _tiered_storage;
+    bool _iceberg;
+    bool _cloud_topics;
+};
+
 class ntp_config {
```

Comment:
Do we need to update `archival_metadata_stm_factory::create()` too?

Reply:
yes